### DISCUSSION TITLE
SDKMan Example

### DISCUSSION THREAD

**Discussion by sualeh** [jreleaser/jreleaser#712] [Q&A]
@aalmiray - do you know of any project that has implemented JReleaser with SDKMan? I am looking for an example. I received a public key for the SDKMan credentials, and I am trying to find out how to obtain and use `JRELEASER_SDKMAN_CONSUMER_KEY` and `JRELEASER_SDKMAN_CONSUMER_TOKEN`.

**Comment by aalmiray** [comment]
You should have received both key & token from the Sdkman crew. At least that used to be the case in the past. @helpermethod my be able to provide more information.

**Reply by sualeh** [reply]
Let me double check. It may be my bad.

**Reply by sualeh** [reply]
@aalmiray - sorry - I have received them. Let me try it out my JReleaser config now.

**Comment by sualeh** [comment]
In my SDKMan packager, I set "downloadUrl" like this:

```yaml
  sdkman:
    active: ALWAYS
    continueOnError: false
    downloadUrl:  https://github.com/schemacrawler/SchemaCrawler/releases/download/v{{projectVersion}}/schemacrawler-{{projectVersion}}-bin.zip
    candidate: schemacrawler
    command: MAJOR
    releaseNotesUrl: https://www.schemacrawler.com/changes-report.html
```

and afterwards, like this:

```yaml
  sdkman:
    active: ALWAYS
    continueOnError: false
    candidate: schemacrawler
    command: MAJOR
    releaseNotesUrl: https://www.schemacrawler.com/changes-report.html
    extraProperties:
      downloadUrl:  https://github.com/schemacrawler/SchemaCrawler/releases/download/v{{projectVersion}}/schemacrawler-{{projectVersion}}-bin.zip
```

In both cases I get an exception:

```
java.lang.NullPointerException: Cannot invoke "String.split(String)" because "coords" is null
	at jreleaser.shadow.org.jreleaser.model.util.Artifacts.resolveDownloadUrlFromUploader(Artifacts.java:383)
	at jreleaser.shadow.org.jreleaser.model.util.Artifacts.resolveDownloadUrl(Artifacts.java:345)
	at jreleaser.shadow.org.jreleaser.model.util.SdkmanHelper.artifactUrl(SdkmanHelper.java:112)
	at jreleaser.shadow.org.jreleaser.model.util.SdkmanHelper.collectArtifacts(SdkmanHelper.java:72)
	at jreleaser.shadow.org.jreleaser.packagers.SdkmanPackagerProcessor.doPublishDistribution(SdkmanPackagerProcessor.java:63)
	at jreleaser.shadow.org.jreleaser.packagers.AbstractPackagerProcessor.publishDistribution(AbstractPackagerProcessor.java:182)
	at jreleaser.shadow.org.jreleaser.packagers.SdkmanPackagerProcessor.publishDistribution(SdkmanPackagerProcessor.java:42)
	at jreleaser.shadow.org.jreleaser.engine.distribution.DistributionProcessor.publishDistribution(DistributionProcessor.java:141)
```

Any suggestions?

**Reply by sualeh** [reply]
Wait - let me see what the uploader needs to be.

**Comment by sualeh** [comment]
@aalmiray - ok, I am officially stuck. I added an uploader and I am getting same the exception. Please take a look at my [jreleaser.yml](https://github.com/schemacrawler/SchemaCrawler-Installers/blob/master/jreleaser.yml).

Ideally, I want my config to look like this:

```yaml
  sdkman:
    active: ALWAYS
    continueOnError: false
    downloadUrl:  https://github.com/schemacrawler/SchemaCrawler/releases/download/v{{projectVersion}}/schemacrawler-{{projectVersion}}-bin.zip
    candidate: schemacrawler
    command: MAJOR
    releaseNotesUrl: https://www.schemacrawler.com/changes-report.html
```

**Reply by aalmiray** [reply]
You have stumbled upon https://github.com/jreleaser/jreleaser/issues/699 which is triggered by setting `skipRelease: true`.

Do not set an HTTP uploader, it's not going to work as you think. That's for a custom HTTP uploader under your control.

**Reply by sualeh** [reply]
Thanks, @aalmiray - that helps.

**Comment by sualeh** [comment]
@aalmiray - I have got a SDKMan release with JReleaser - if anyone needs a code example, it is at https://github.com/schemacrawler/SchemaCrawler-Installers.

One last question remains:  the SDKMan packager runs on both Windows and Linux. Should it run only on Linux and not Windows? If I have runners on both platforms, how can I selectively disable on one platform?

**Reply by aalmiray** [reply]
Actually Sdkman runs on macOs as well. There's no need to restrict downloads for a particular platform, rather do the opposite, enable more!

**Reply by aalmiray** [reply]
Which also makes me wonder why there's no macOs binary for SchemaCrawler that bundles a Java Runtime with jpackage, like it's done for Linux and Windows.

**Reply by sualeh** [reply]
> Actually Sdkman runs on macOs as well. There's no need to restrict downloads for a particular platform, rather do the opposite, enable more!

@aalmiray - I have one JReleaser script that has packagers for Chocolatey and SDKMan. I run that on a Windows GitHub Actions runner, as well as Ubuntu runner. The second SDKMan packager to run fails, since the packager will run on either platform.  The Chocolatey packager does not fail, since it only runs on Windows. How do I specify that the SDKMan packager should only run on one of the platforms?

**Reply by sualeh** [reply]
> Which also makes me wonder why there's no macOs binary for SchemaCrawler that bundles a Java Runtime with jpackage, like it's done for Linux and Windows.

Mainly because I do not have a Mac to test the output on. 😄

**Reply by aalmiray** [reply]
> @aalmiray - I have one JReleaser script that has packagers for Chocolatey and SDKMan. I run that on a Windows GitHub Actions runner, as well as Ubuntu runner. The second SDKMan packager to run fails, since the packager will run on either platform.  The Chocolatey packager does not fail, since it only runs on Windows. How do I specify that the SDKMan packager should only run on one of the platforms?

If you're running the release/package steps multiple times you may include or exclude packagers, distributions, announcers, uploaders.

For example you may `jrelease publish -p sdkman` which will only activate sdkman for all distributions, or you may `jreleaser publish -xp chocolatey` which will run all configured packagers except for Chocolatey.

**Reply by aalmiray** [reply]
> Mainly because I do not have a Mac to test the output on. 😄

No worries. I don't use Windows yet JReleaser produces binaries for it 😉

**Reply by sualeh** [reply]
> > Mainly because I do not have a Mac to test the output on. 😄
> 
> No worries. I don't use Windows yet JReleaser produces binaries for it 😉

You are braver than me. 😄

**Reply by sualeh** [reply]
> > @aalmiray - I have one JReleaser script that has packagers for Chocolatey and SDKMan. I run that on a Windows GitHub Actions runner, as well as Ubuntu runner. The second SDKMan packager to run fails, since the packager will run on either platform.  The Chocolatey packager does not fail, since it only runs on Windows. How do I specify that the SDKMan packager should only run on one of the platforms?
> 
> If you're running the release/package steps multiple times you may include or exclude packagers, distributions, announcers, uploaders.
> 
> For example you may `jrelease publish -p sdkman` which will only activate sdkman for all distributions, or you may `jreleaser publish -xp chocolatey` which will run all configured packagers except for Chocolatey.

Thanks. For now, since I have only set up SDKMan and Chocolatey, I will run just on a Windows runner. I will revisit that when I add more packagers.
